meteor/lib/Rundown.ts

```diff
@@ -314,6 +315,61 @@ export function getSegmentsWithPartInstances(
 	})
 }
 
+export function getUnfinishedPieceInstances(
+	playlistActivationId: RundownPlaylistActivationId,
+	currentPartInstanceId: PartInstanceId,
+	showStyleBase: ShowStyleBase,
+	invalidateAt?: Function
+) {
+	const now = getCurrentTime()
+	let prospectivePieces: PieceInstance[] = []
+
+	const partInstance = PartInstances.findOne(currentPartInstanceId)
+
+	if (partInstance) {
+		prospectivePieces = PieceInstances.find({
+			partInstanceId: currentPartInstanceId,
+			playlistActivationId: playlistActivationId,
+		}).fetch()
+
+		const nowInPart = partInstance.timings?.startedPlayback ? now - partInstance.timings.startedPlayback : 0
+		prospectivePieces = processAndPrunePieceInstanceTimings(showStyleBase, prospectivePieces, nowInPart)
+
+		let nearestEnd = Number.POSITIVE_INFINITY
+		prospectivePieces = prospectivePieces.filter((pieceInstance) => {
+			const piece = pieceInstance.piece
+
+			if (!pieceInstance.adLibSourceId && !piece.tags) {
+				// No effect on the data, so ignore
+				return false
+			}
+
+			let end: number | undefined
+			if (pieceInstance.stoppedPlayback) {
+				end = pieceInstance.stoppedPlayback
+			} else if (pieceInstance.userDuration && typeof pieceInstance.userDuration.end === 'number') {
+				end = pieceInstance.userDuration.end
+			} else if (typeof piece.enable.duration === 'number' && pieceInstance.startedPlayback) {
+				end = piece.enable.duration + pieceInstance.startedPlayback
+			}
+
+			if (end !== undefined) {
+				if (end > now) {
+					nearestEnd = Math.min(nearestEnd, end)
+					return true
+				} else {
+					return false
+				}
+			}
+			return true
```

Comment:
To reduce the nesting a bit, this can be rewritten to something like:
```suggestion
			if (end === undefined) {
				return true
			}
			if (now <= end) {
				return false
			}
			nearestEnd = Math.min(nearestEnd, end)
			return true
```